### PR TITLE
Test N+1 in /buyers/accounts index

### PR DESCRIPTION
**Pull request by jlledom**
**What this PR does / why we need it**:

Adds a test to check N+1 is solved in the `/buyers/accounts` screen

**Which issue(s) this PR fixes** 

[THREESCALE-8530](https://issues.redhat.com/browse/THREESCALE-8530)

**Special notes for your reviewer**:

Not sure if I did this correctly, I filtered by table `users` and requested the `json` format to reduce  the number of observed queries. Also set the `slope` to 3 because that's the number of queries for each account that I observed in previous tests.

### CODE REVIEW

File: test/integration/buyers/accounts_controller_test.rb [hunk 1]
```diff
@@ -391,4 +391,18 @@ class NotLoggedInTest < ActionDispatch::IntegrationTest
...
+
+  class ManyAccountsTest < ActionDispatch::IntegrationTest
+    def setup
+      @provider = FactoryBot.create(:provider_account)
+      login! @provider
+    end
+
+    test 'no N+1 problems in the index' do
+      populate = ->(n) { FactoryBot.create_list(:buyer_account, n, provider_account: @provider) }
+      assert_perform_linear_number_of_queries(slope: 3, populate: populate, matching: /users/) do
```
Comment (akostadinov) [inline test/integration/buyers/accounts_controller_test.rb:403]: Shouldn't we `assert_perform_constant_number_of_queries`? Why aren't all buyers queried by a single query?

Just asking, I didn't look at the code. Just this is how I would expect endpoint to work.

**Comment by akostadinov** [comment]
While on it, could you include JIRA issue in commit comment? This helps when looking at git history without need to go to and rely on github.

File: test/integration/buyers/accounts_controller_test.rb (hunk 1 above)
Comment (akostadinov) [reply test/integration/buyers/accounts_controller_test.rb:403]: btw in the past I had to use `warmup` to avoid issues caused by caching or who knows what.

Comment (jlledom) [reply test/integration/buyers/accounts_controller_test.rb:403]: I think you're right and I misunderstood this. But I don't understand why the amount of queries to the users table keeps incrementing linearly after solving the N+1 and bullet stopped complaining. It's possible that there are more N+1 problems in this screen not detected by bullet?. I'll investigate further.

**Comment by jlledom** [comment]
I found the problem here. There was another N+1 in that screen, not detected by Bullet, at this line:

[app/views/buyers/accounts/_search_results.html.slim#L58](https://github.com/3scale/porta/blob/0e946cc325b60719fc3eff9820174b39da278765/app/views/buyers/accounts/_search_results.html.slim#L58)

Accessing `first_admin` filters by scopes `admin` and `but_impersonation_admin`. Since the scopes are defined as a `where` clause, a new SQL query is executed every time `first_admin` is called.

To fix this, I modified the account decorator to apply the filter manually when the users association happens to be already loaded, avoiding the extra query, and then including the association in the main query in the account index presenter.

File: app/models/user.rb [hunk 2]
```diff
@@ -140,7 +140,7 @@ def self.search_states
 
   scope :latest, -> {limit(5).order(created_at: :desc)}
 
-  scope :but_impersonation_admin, -> { where(['username <> ?', ThreeScale.config.impersonation_admin['username']]) }
+  scope :but_impersonation_admin, -> { where(['users.username <> ?', ThreeScale.config.impersonation_admin['username']]) }
```
Comment (akostadinov) [inline app/models/user.rb:143]: why is this needed?

File: app/decorators/account_decorator.rb [hunk 3]
```diff
@@ -3,8 +3,24 @@
...
+  def first_admin_cached
+    @first_admin_cached ||= admins&.first
+  end
+
+  def admins
+    return nil unless users.loaded?
+
+    @admins ||= users.select do |user|
+      user.role == :admin && user.username != ThreeScale.config.impersonation_admin['username']
+    end
```
Comment (akostadinov) [inline app/decorators/account_decorator.rb:21]: maybe add a method `#admin?` for the `Account`

File: app/decorators/account_decorator.rb [hunk 4]
```diff
@@ -3,8 +3,24 @@
 class AccountDecorator < ApplicationDecorator
   delegate :display_name, :email, to: :admin_user, prefix: true
 
+  def first_admin
+    @first_admin ||= (first_admin_cached || super)
+  end
```
Comment (akostadinov) [inline app/decorators/account_decorator.rb:8]:
```suggestion
  def first_admin
    @first_admin ||= users.loaded? ? users.find { |user| user.admin? && user.impersonation_admin? }  : super
  end
```

Where `user.impersonation_admin?` needs to be implemented in `User`.

**Review by akostadinov — CHANGES_REQUESTED** [review]

File: app/models/user.rb (hunk 2 above)
Comment (jlledom) [reply app/models/user.rb:143]: I found some tests failing because username was an ambiguous column name: https://app.circleci.com/pipelines/github/3scale/porta/21537/workflows/052cd6c3-150c-44f3-a943-a83579956da4/jobs/249337

File: app/decorators/account_decorator.rb (hunk 4 above)
Comment (jlledom) [reply app/decorators/account_decorator.rb]: I used this code, and I also esu the `:admin_users` association instead of `:users`, since `first_admin` only searches among admins.

File: app/models/user.rb (hunk 2 above)
Comment (akostadinov) [reply app/models/user.rb:143]:
```suggestion
  scope :but_impersonation_admin, -> { where.has { username <> ThreeScale.config.impersonation_admin['username'] }}
```
What I'm concerned with is hardcoding table name might be a problem with more complicated joins or sub-queries. I'm not sure that's a real risk but this is my reasoning :man_shrugging:

File: app/presenters/buyers/accounts_index_presenter.rb [hunk 5]
```diff
@@ -14,7 +14,7 @@ def initialize(provider:, params: {})
 
   def buyers
     @buyers ||= provider.buyer_accounts
-                        .includes([:admin_user])
+                        .includes(%i[admin_users admin_user])
```
Comment (akostadinov) [inline app/presenters/buyers/accounts_index_presenter.rb:17]: why is `admin_user` not enough?

File: app/models/user.rb (hunk 2 above)
Comment (jlledom) [reply app/models/user.rb]: done

File: app/presenters/buyers/accounts_index_presenter.rb (hunk 5 above)
Comment (jlledom) [reply app/presenters/buyers/accounts_index_presenter.rb]: I think it's enough, and all the first_admin additional logic was redundant, I removed it and the n+1 test passes

File: app/presenters/buyers/accounts_index_presenter.rb [hunk 6]
```diff
@@ -14,7 +14,7 @@ def initialize(provider:, params: {})
 
   def buyers
     @buyers ||= provider.buyer_accounts
-                        .includes([:admin_user])
+                        .includes(%i[admin_user])
```
Comment (akostadinov) [inline app/presenters/buyers/accounts_index_presenter.rb:17]: Everything looks awesome now. Just a nitpick. I prefer such minor style changes to stay out because they make git blame more complicated. I don't know whether you use some special tool to help with that.

File: test/integration/buyers/accounts_controller_test.rb [hunk 7]
```diff
@@ -391,4 +391,18 @@ class NotLoggedInTest < ActionDispatch::IntegrationTest
...
+
+  class ManyAccountsTest < ActionDispatch::IntegrationTest
+    def setup
+      @provider = FactoryBot.create(:provider_account)
+      login! @provider
+    end
+
+    test 'no N+1 problems in the index' do
+      populate = ->(n) {FactoryBot.create_list(:buyer_account, n, provider_account: @provider)}
+      assert_perform_constant_number_of_queries(populate: populate, matching: /users/) do
```
Comment (akostadinov) [inline test/integration/buyers/accounts_controller_test.rb:403]: do we need the `matching` argument?

Comment (jlledom) [reply test/integration/buyers/accounts_controller_test.rb:403]: I think so, I'm using it because the N+1 problem was querying the users table. If I remove it, then it fails b/c the number of queries to the plans table also increments linearly

**Review by akostadinov — APPROVED** [review]
awesome work!

please squash when you decide to merge
